### ISSUE TITLE
Multi-layer perceptron

### ISSUE THREAD

**Issue by marfox** [Wikidata/soweego#286]
- [x] Implement a `MultiLayerPerceptron` class in `soweego.linker.neural_networks`;
- [x] investigate the best model architecture, i.e., # of layers, dropouts, etc.;
- [x] pick the best loss
- [x] pick best activation function (`relu` seems to work best)
- [x] pick best optimization function (`adadelta` performs best)
- [x] apply `mlp` to all `imdb` entities
- [x] apply `mlp` to all `discogs` entities
- [x] apply `mlp` to all `musicbrainz` entities

**Comment by tupini07** [comment]
I've tried different architectures and they all seem to have a similar performance. Below are the results of the experiments in list form. These are the results of applying the evaluation of the `mlp` (multi layer perceptron) on _IMDb Writer_. The command to do this is:

`python -m soweego linker evaluate mlp imdb writer` 

_Note: 'BN' stands for 'Batch Normalization' layer. All other layers are 'Dense' layers._

## 64 > BN > 1


- Precision:
    - mean = 0.9177771662962314
    - standard deviation = 0.005186747481101013
- Recall:
    - mean = 0.9302842678950632
    - standard deviation = 0.0036075816954025783
- F-score:
    - mean = 0.9239773239577866
    - standard deviation = 0.0031370601795963366


## 128 > BN > 1


- Precision:
    - mean = 0.9202362134964461
    - standard deviation = 0.0032664858937136603
- Recall:
    - mean = 0.9292820545047201
    - standard deviation = 0.0030519660791793256
- F-score:
    - mean = 0.9247340663992976
    - standard deviation = 0.0026945762929905816



## 256 > BN > 1


- Precision:
    - mean = 0.9162807274632913
    - standard deviation = 0.003924488061212346
- Recall:
    - mean = 0.9329570804029442
    - standard deviation = 0.0022279075274187444
- F-score:
    - mean = 0.9245344414506821
    - standard deviation = 0.001329073074095115


## 64 > BN > 32 > BN > 1


- Precision:
    - mean = 0.9162443955390298
    - standard deviation = 0.002545288536287673
- Recall:
    - mean = 0.9342935280003319
    - standard deviation = 0.0030050798754369897
- F-score:
    - mean = 0.9251744573032659
    - standard deviation = 0.0013116538151647197


## 128 > BN > 32 > BN > 1


- Precision:
    - mean = 0.9159962774642854
    - standard deviation = 0.0032426708628756455
- Recall:
    - mean = 0.9355931937110002
    - standard deviation = 0.003970157051690947
- F-score:
    - mean = 0.9256779301551065
    - standard deviation = 0.000959995764984693


## 32 > BN > 32 > BN > 32 > BN > 1


- Precision:
    - mean = 0.9153877339268387
    - standard deviation = 0.004302163619745764
- Recall:
    - mean = 0.9330316639814097
    - standard deviation = 0.008785712504418706
- F-score:
    - mean = 0.9240801068136706
    - standard deviation = 0.0022677049354925807


## 64 > BN > 32 > BN > 32 > BN > 1


- Precision:
    - mean = 0.9151180938045229
    - standard deviation = 0.003796321736802397
- Recall:
    - mean = 0.9356672673869524
    - standard deviation = 0.006000688618184766
- F-score:
    - mean = 0.925254418270909
    - standard deviation = 0.0015799854229909295




## 32 > BN > 32 > BN > 32 > BN > 32 > BN > 1


- Precision:
    - mean = 0.917488064665403
    - standard deviation = 0.005321593373539414
- Recall:
    - mean = 0.9314350558529298
    - standard deviation = 0.00421042075304083
- F-score:
    - mean = 0.9243865236364496
    - standard deviation = 0.0015486946126652362



## 64 > BN > 32 > BN > 32 > BN > 32 > BN > 1


- Precision:
    - mean = 0.9192112562285286
    - standard deviation = 0.007715791858829185
- Recall:
    - mean = 0.9288735330139138
    - standard deviation = 0.008821425468741293
- F-score:
    - mean = 0.9239471584688012
    - standard deviation = 0.0019287451377498477



## 128 > BN > 32 > BN > 32 > BN > 32 > BN > 1


- Precision:
    - mean = 0.9170497101707097
    - standard deviation = 0.005438362287916304
- Recall:
    - mean = 0.9311754534583725
    - standard deviation = 0.007938220196513236
- F-score:
    - mean = 0.9240150511252165
    - standard deviation = 0.0023696848352501316



## 128 > BN > 64 > BN > 32 > BN > 32 > BN > 1


- Precision:
    - mean = 0.919650582662104
    - standard deviation = 0.0038994435812963768
- Recall:
    - mean = 0.9281683791811488
    - standard deviation = 0.0033804371044485852
- F-score:
    - mean = 0.9238825164019586
    - standard deviation = 0.0025695692789418547


## 128 > BN > 256 > BN > 64 > BN > 32 > BN > 1


- Precision:
    - mean = 0.9210188492991502
    - standard deviation = 0.004061779578133305
- Recall:
    - mean = 0.9267582437799817
    - standard deviation = 0.005851255653502437
- F-score:
    - mean = 0.9238552214429074
    - standard deviation = 0.0016341446276613744


## 128 > BN > 64 > BN > 32 > BN > 32 > BN > 16 > BN > 1


- Precision:
    - mean = 0.9141589182625489
    - standard deviation = 0.008947959668081396
- Recall:
    - mean = 0.9347765503827092
    - standard deviation = 0.0065009726349377185
- F-score:
    - mean = 0.9242887408585239
    - standard deviation = 0.0016748753612694514





## 32 > BN > 32 > BN > 32 > BN > 32 > BN > 32 > BN > 32 > BN > 32 > BN > 1


- Precision:
    - mean = 0.9203865256668212
    - standard deviation = 0.004457127081998804
- Recall:
    - mean = 0.9259779964662378
    - standard deviation = 0.005135325978060001
- F-score:
    - mean = 0.9231491718135757
    - standard deviation = 0.0006258036311244564

**Comment by tupini07** [comment]
This is a table presenting the same results listed in the previous comment, but in an easier to read tabular format. The bolded numbers represent the best score obtained for that specific column. 


| Model                                                                   | Precision (mean) | Precision (st.dev) | Recall (mean) | Recall (st.dev) | F-score (mean) | F-score (st.dev) |
|-------------------------------------------------------------------------|------------------|--------------------|---------------|-----------------|----------------|------------------|
| _64 > BN > 1_                                                             | 0.9177           | 0.0051             | 0.9302        | 0.0036          | 0.9239         | 0.0031           |
| _128 > BN > 1_                                                            | 0.9202           | 0.0032             | 0.9292        | 0.0030          | 0.9247         | 0.0026           |
| _256 > BN > 1_                                                            | 0.9162           | 0.0039             | 0.9329        | **0.0022**          | 0.9245         | 0.0013           |
| _64 > BN > 32 > BN > 1_                                                   | 0.9162           | **0.0025**             | 0.9342        | 0.0030          | 0.9251         | 0.0013           |
| _128 > BN > 32 > BN > 1_                                                  | 0.9159           | 0.0032             | 0.9355        | 0.0039          | **0.9256**         | **0.0009**           |
| _32 > BN > 32 > BN > 32 > BN > 1_                                         | 0.9153           | 0.0043             | 0.9330        | 0.0087          | 0.9240         | 0.0022           |
| _64 > BN > 32 > BN > 32 > BN > 1_                                         | 0.9151           | 0.0037             | **0.9356**        | 0.0060          | 0.9252         | 0.0015           |
| _32 > BN > 32 > BN > 32 > BN > 32 > BN > 1_                               | 0.9174           | 0.0053             | 0.9314        | 0.0042          | 0.9243         | 0.0015           |
| _64 > BN > 32 > BN > 32 > BN > 32 > BN > 1_                               | 0.9192           | 0.0077             | 0.9288        | 0.0088          | 0.9239         | 0.0019           |
| _128 > BN > 32 > BN > 32 > BN > 32 > BN > 1_                              | 0.9170           | 0.0054             | 0.9311        | 0.0079          | 0.9240         | 0.0023           |
| _128 > BN > 64 > BN > 32 > BN > 32 > BN > 1_                              | 0.9196           | 0.0038             | 0.9281        | 0.0033          | 0.9238         | 0.0025           |
| _128 > BN > 256 > BN > 64 > BN > 32 > BN > 1_                             | **0.9210**           | 0.0040             | 0.9267        | 0.0058          | 0.9238         | 0.0016           |
| _128 > BN > 64 > BN > 32 > BN > 32 > BN > 16 > BN > 1_                    | 0.9141           | 0.0089             | 0.9347        | 0.0065          | 0.9242         | 0.0016           |
| _32 > BN > 32 > BN > 32 > BN > 32 > BN > 32 > BN > 32 > BN > 32 > BN > 1_ | 0.9203           | 0.0044             | 0.9259        | 0.0051          | 0.9231         | 0.0006           |

**Comment by tupini07** [comment]
By referencing the above table we can infer that a smaller model, but not too small, performs best. At first glance it seems that the `128 > BN > 32 > BN > 1` architecture would be the best one to use for the `mlp` classifier.

However, by comparing the obtained results with those presented for `imdb/writer` in the [Linkers evaluation](https://github.com/Wikidata/soweego/wiki/Linkers-evaluation#imdb-writer-1) page we can see that at least for the main metrics (_precision_, _recall_, and _F1_) `mlp` does not provide a significant improvement (if any) over the already existing classifiers. In fact, `slp` seems to outperform `mlp` in every metric except for the `recall`, where `mlp` surpases `slp` by `0.0075`. 

It would still be worthwhile to see how the `mlp` classifier evaluates when applied to other entities, especially those from other catalogs.

**Comment by marfox** [comment]
@tupini07 , I ran evaluation, training, and classification over all entities of all catalogs. You can find the results at `soweego-2.eqiad.wmflabs:/srv/dev/20190503`.
Can you please update the [Wiki page](https://github.com/Wikidata/soweego/wiki/Linkers-evaluation#comparison) accordingly?
Then feel free to close this task.

**Comment by tupini07** [comment]
Sure! Wiki page has been updated. Closing this issue
